Runtime: adding more validation to workflow

SatyamMattoo opened the pull request:
## Short Description

Added functions to validate the workflow, steps and options and return errors or warnings accordingly

## Related issue

Fixes #681 

## Implementation Details

Added more validation functions to `validate-plan.ts` returning errors or warnings for a bad workflow structure.

## QA Notes

The following approach is not breaking any existing test cases. But we can add new test
cases where we can check the warning logs are printed successfully

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have added unit tests
- [ ] Changesets have been added (if there are production code changes)

josephjclark commented:
Hi @SatyamMattoo , just to check, this and other issues aren't covered by the DMP programme. This has no relation to c4gt and does not influence  any DMP application you make.

If you want to take on a project under DMP - and I hope you will! - you need to apply directly through the C4GT portal. I don't know anything about that, I'm not involved in the process at all.

SatyamMattoo commented:
Hey @josephjclark 
Sorry for the misunderstanding, I forgot to change the branch name here because I was working on a different issue as well. I have applied for the DMP program as well and this isn't related to that. You can have a look and suggest any changes that might be needed here. 

Best regards!

josephjclark commented on packages/runtime/src/util/validate-plan.ts:
```diff
@@ -20,6 +24,54 @@ export default (plan: ExecutionPlan) => {
+  });
+
+  if (options) {
+    assertOptionsStructure(options);
```
You can remove this if statement if you default `options` to an empty object inside `assertOptionStructure`, resulting in more elegant code :)

```
const assertOptionsStructure = (options: WorkflowOptions = {}) => { ... }
```

josephjclark commented on packages/runtime/src/util/validate-plan.ts:
```diff
@@ -20,6 +24,54 @@ export default (plan: ExecutionPlan) => {
+const assertOptionsStructure = (options: WorkflowOptions) => {
+  const allowedKeys = ['timeout', 'stepTimeout', 'start', 'end', 'sanitize'];
+
+  Object.keys(options).forEach((key) => {
```
Try using `for in`:

```
for (const key in options ) {
  if (!allowedKeys.includes(key)) {
      logger.warn(`Warning: Unrecognized option "${key}" in options object.`);
    }
}
```

josephjclark commented on packages/runtime/src/util/validate-plan.ts:
```diff
@@ -20,6 +24,54 @@ export default (plan: ExecutionPlan) => {
+
+  Object.keys(options).forEach((key) => {
+    if (!allowedKeys.includes(key)) {
+      logger.warn(`Warning: Unrecognized option "${key}" in options object.`);
```
Please don't include full stops after log messages. They're unusual and unnecessary in log output

josephjclark reviewed (changes requested):
Hi @SatyamMattoo 

I've left a few comments on particular details.

But after thinking about it for a few days, I really think this level of validation needs to move into the `cli` package. Having these kind of heavyweight validation steps in the runtime is just uneccessary

In `packages/cli/src/execute/handler.ts` there is a call to `loadPlan` inside `executeHandler. I would add a call to `validatePlan` after this, and copy your validation rules into new file in `cli/src/util/`

The existing validation in the runtime, which validates the logical consistency of the execution plan, needs to stay where it is. But these new structural/syntactic validations need to move out and into the CLI.

Sorry for the inconvenience - this is on my for specifying bad requirements in the first place. I just didn't know until I saw the code.

SatyamMattoo commented on packages/runtime/src/util/validate-plan.ts:
```diff
@@ -1,16 +1,20 @@
   down: Record<string, true>;
 };
 
+const logger = createLogger(undefined, { json: true });
```
Do we need to create a logger in the main function and then pass it to the functions like 
`const assertWorkflowStructure = (plan: ExecutionPlan, logger: Logger)` ?

Additionally, should we assign a name to the logger when creating it?
```
export default (plan: ExecutionPlan) => {
  const logger = createLogger(undefined, { json: true });
  assertWorkflowStructure(plan, logger);
```

SatyamMattoo commented:
Hello @josephjclark,

I have made the requested changes. Please review them at your convenience and let me know if there are any additional modifications needed.

Best regards.

josephjclark commented:
Thank you @SatyamMattoo - I'll get back to you next week when I have time :)

josephjclark commented:
Hi @SatyamMattoo actually I've found a spare 20 minutes so  I'm taking a look now.

I don't think you've seem my [main comment](https://github.com/OpenFn/kit/pull/696#pullrequestreview-2070233256) on the last review. We need to move these new validation steps out of the runtime and into the CLI.

SatyamMattoo commented:
Hi @josephjclark I have moved the validation from the runtime to the CLI and also added a few more tests for the logs. You can have a look at your convenience and let me know if there are any more changes needed.

Best regards.

josephjclark commented on packages/cli/test/util/validate-plan.test.ts:
```diff
@@ -0,0 +1,118 @@
+})
+
+test('throws for missing workflow', (t) => {
+    //@ts-ignore
```
These `@ts-ignore`s are pretty awkward.

What I usually do is force a type assertion:
```
 const plan = {
        options: {
            start: 'a',
        }
    } as ExecutionPlan;
```

This usually does the trick. Sometimes you need to hack it with `as unknown as ExecutionPlan`

josephjclark commented on packages/cli/test/util/validate-plan.test.ts:
```diff
@@ -0,0 +1,118 @@
+    });
+});
+
+test('throws for adaptor without an expression', (t) => {
```
This title should be: `throws for a step with an adaptor but no expression`

SatyamMattoo commented on packages/cli/test/util/validate-plan.test.ts:
```diff
@@ -0,0 +1,118 @@
+    };
+
+    t.throws(() => validate(plan, logger), {
+        message: 'Missing or invalid workflow key in execution plan',
```
This error message was actually related to the 'workflow' key, so I have changed the message to 
`throw new Error("Missing or invalid 'workflow' key in execution plan");`

josephjclark replied:
Ah I see! It would be better english to say `Missing or invalid key "workflow" in execution plan` (double quotes is deliberate)

SatyamMattoo replied:
Done!

SatyamMattoo commented:
Hey @josephjclark, I have committed the requested changes. You can review them at your convenience and let me know if there are any further changes required.

Sorry for the late response, I was occupied with my exams recently.

Best regards.

josephjclark commented:
Thank you @SatyamMattoo - this will be released shortly!
